docs/developers/deprecations.md

```diff
@@ -0,0 +1,67 @@
+# Deprecations
+
+✋ Before reading this, please read the
+[API Compatibility Policy](../../api_compatibility_policy.md)
+
+The following are things developers should keep in mind for deprecations. These
+tips may not apply in every case, so use your best judgement! If you're not sure
+how these affect your deprecation or timeline, ask the maintainers.
+
+1. Assume default values are used.
+
+   Feature flags are an excellent way for users to opt-in and try out new
+   behavior before it goes live. However unless someone is trying to experiment
+   with the new feature, many users and other tools built on top of the API
+   won't know about it until it becomes the new default.
+
+   Try to give users as much time with a new feature flag in opt-out mode as you
+   can. Users may not update releases immediately, so this gives you more time
+   to get feedback from a broader set of users while still having an escape
+   hatch to revert back to old behavior if problems come up. This may mean
+   splitting up your deprecation window into opt-in and opt-out phases. For
+   particularly disruptive changes, consider making the switch to opt-out its
+   own deprecation notification.
+
+   A conservative deprecation sequence to move from field A -> B might look
+   like:
+
+   1. (inital state) Return field A
+   2. Introduce new field B with opt-in feature flag (disabled by default).
+   3. Set feature flag to return both A and B in responses by default.
+   4. Set feature flag to opt-in users by default and only return field B.
+   5. Remove feature flag.
```

Comment:
to clarify, are all these steps within the 9 months deprecation period? just want to be clear that we won't double the deprecation time

Reply:
There's no one right answer here - this is going to depend on what the change is, how the deprecation is done, and how disruptive it is.

You can make the argument that both the change in default behavior of a feature flag and the removal of a feature flag should be separate deprecation events. This is especially true for any deprecation that results in xor behavior.

[I think there is room for shortcuts](https://github.com/tektoncd/pipeline/pull/6208#discussion_r1114719628), so long as we have reasonable confidence that clients will be able to notice and adapt to changes - but I think how this can be done will differ on a case by case basis.